### DISCUSSION TITLE
Frontend trust relationship changes when hostname-admin is set

### DISCUSSION THREAD

**Discussion by baparker** [keycloak/keycloak#34523] [Q&A] [answered]
On Keycloak 26.0.2 I'm trying to use the hostname-admin configuration option by setting it to an internal address, but I ran into a trust relationship issue when making admin API calls using the frontend URL which is causing failures in my setup scripts.

The admin console works fine and I can access it with both URLs, but I'm getting trust relationship issues for the SSL/TLS secure channel when I make admin API calls using the frontend URL. Without the hostname-admin option, my admin API calls are trusted (the certificate I'm using has the frontend URL so it should be trusted).

From the Hostname (v2) documentation, "Using the hostname-admin option does not prevent accessing the Administration REST API endpoints via the frontend URL specified by the hostname option. [...] Administration Console implicitly accesses the API using the URL as specified by the hostname-admin option.". But it seems like all calls to the admin APIs (not just the admin console calls to the admin API) are implicitly using the hostname-admin URL, even when the frontend URL is used to make the API call.

Why would my admin API calls using the frontend URL no longer be trusted when hostname-admin is set? I'm confused because the URL matches the URL in my certificate. Maybe this is a bug?

**Comment by shawkins** [comment]
When you say no longer trusted, what exception and/or response code are you getting back?

**Reply by baparker** [reply]
![image](https://github.com/user-attachments/assets/2475ee48-1951-4d42-b12b-e26c760991fb)
This is from PowerShell's Invoke-RestMethod

**Reply by shawkins** [reply]
Are you able to get debug output from PowerShell? There has to be a reason why it doesn't like the server certificate. Based upon the description you are giving I would assume that hostname validation is failing. 

Is the server certificate you have configured valid for the hostname, but not the hostname-admin? 

Can you describe what the powershell logic is doing so that we can also determine why it may be using hostname-admin - for example via a redirect or some returned url from Keycloak.

**Reply by baparker** [reply]
> Are you able to get debug output from PowerShell? There has to be a reason why it doesn't like the server certificate. Based upon the description you are giving I would assume that hostname validation is failing.

The error is not very specific, but it seems like the hostname validation is failing to me too. Most detail I could get:

![image](https://github.com/user-attachments/assets/e83559a4-c798-444a-94e2-3b00546e027a)

> Is the server certificate you have configured valid for the hostname, but not the hostname-admin?

Yes, that's exactly right.

> Can you describe what the powershell logic is doing so that we can also determine why it may be using hostname-admin - for example via a redirect or some returned url from Keycloak.

I started trying out the various API calls I'm making in my script manually, and I realized the trust relationship fails using just the frontend URL e.g. https://fqdn:port. The other calls are actually working e.g. frontendURL/admin/realms. I was using the frontend URL to ping Keycloak so I knew it was running. I replaced that call in my script with a different one, and all the calls in my script are working now. 
I originally chose that URL because it did not require credentials for a success. I tried frontendURL/admin/master/console and that is trusted and does not require credentials, so I'll use that instead.

Something odd is still going on though. It seems the trust relationship is changed only for the frontendURL when hostname-admin is configured. But when I login to the admin console using the browser, I do not need to accept the 'invalid' certificate whether hostname-admin is set or not- so everything is trusted via the browser. I'm really not sure why the frontend URL is no longer trusted when hostname-admin is set, while the admin console via the browser and the admin APIs are trusted.

**Reply by shawkins** [reply, marked as answer]
> I started trying out the various API calls I'm making in my script manually, and I realized the trust relationship fails using just the frontend URL e.g. https://fqdn:port/

Trying to reach just https://fqdn:port/ will by default redirect to the admin console for the master realm, which when the hostname-admin is set will be hostname-admin/admin/master/console.

When you directly accesss frontendURL/admin/master/console there is no redirect.
